c10/core/impl/cow/README.md

```diff
@@ -0,0 +1,152 @@
+because users could be relying on the aliasing behavior, intentionally
+or not.
+
+For one release, rather than returning copy-on-write tensors, we
+instead warn when users have triggered behavior in their program that
+relies on the aliasing of the output and input.
+
+To do this, we must simulate the behavior in a backward compatible
+way. To remain backward compatible, the aliases must preserve the
+invariant that they have the same storage. This is a big deviation
+from the design detailed above. To get closer to the real design and
+implementation, we introduce a new `c10::TensorImpl` level concept
+called "Shadow Storage". The shadow storage represents what the
+storage would have looked like the view actually been a lazy copy.
+
+In the instrumented world we thus maintain the formal invariant that
+tensors that alias share a storage. But we have a new invariant:
+tensors that are lazy-copies of each other will share a shadow
+storage.
+
+So what do we warn on? We warn if there is a write to a tensor in a
+set that shares a shadow storage followed by a read or a write to a
+different tensor that shares a physical storage but has a different
+shadow storage. In the real implementation, the first write would have
+triggered a copy, forever cleaving the two sets of tensors, but in the
+current world we instead had behavior that relied on the view-ness of
+the output of reshape.
+
+We can track these violations simply by adding a generation number to
+the shadow and physical storages, updating them both on writes and
+observing if a read or write ever encounters values that are out of
+sync.
+
+We have a few mechanisms for tracking reads and writes:
+ * reads can be tracked by const accesses to the data pointer
+ * writes can be tracked by mutable accesses to the data pointer
+ * writes may also be tracked via autograd, using the same mechanism
+   to bump version numbers
+
+Note that we presently are only checking via autograd, since we don't
+have const access to the data pointer, so we would be way too
+aggressive if we assumed every access was a real write.
+
+### Optimizations to the instrumentation
+Technically, every tensor will require a shadow storage, because if we
+were to create a view of a tensor and then create a lazy-copy, both
+the original tensor and the view would have to share the shadow
+storage, and thus it has to be created and shared before we ever even
+know we needed it.
+
+But we don't want to pay the memory cost of this since we don't expect
+it to be that common. We can get around this by saving the original
+shadow storage on the phsyical storage itself. We will have an
+asymmetrical rule that states that any tensor that has a null shadow
+storage will instead get its shadow storage from the physical
+storage. This allows us to avoid refcount bumps on the shadow storage
+as well as deferring any generation number bumps until we actually
+have an outstanding copy on write.
```

Comment:
This seems slightly out of date now

Reply:
I re-read the whole thing, I don't see what is out of date.